server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java

```diff
@@ -73,554 +49,125 @@
-            this.coerce = gsfm.coerce;
-        }
-        if (gsfm.ignoreMalformed.explicit()) {
-            this.ignoreMalformed = gsfm.ignoreMalformed;
-        }
-        if (gsfm.ignoreZValue.explicit()) {
-            this.ignoreZValue = gsfm.ignoreZValue;
-        }
-    }
-
-    @Override
-    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
-        builder.field("type", contentType());
-
-        if (includeDefaults || fieldType().tree().equals(Defaults.TREE) == false) {
-            builder.field(Names.TREE, fieldType().tree());
-        }
-
-        if (fieldType().treeLevels() != 0) {
-            builder.field(Names.TREE_LEVELS, fieldType().treeLevels());
-        } else if(includeDefaults && fieldType().precisionInMeters() == -1) { // defaults only make sense if precision is not specified
-            if ("geohash".equals(fieldType().tree())) {
-                builder.field(Names.TREE_LEVELS, Defaults.GEOHASH_LEVELS);
-            } else if ("legacyquadtree".equals(fieldType().tree())) {
-                builder.field(Names.TREE_LEVELS, Defaults.QUADTREE_LEVELS);
-            } else if ("quadtree".equals(fieldType().tree())) {
-                builder.field(Names.TREE_LEVELS, Defaults.QUADTREE_LEVELS);
-            } else {
-                throw new IllegalArgumentException("Unknown prefix tree type [" + fieldType().tree() + "]");
-            }
-        }
-        if (fieldType().precisionInMeters() != -1) {
-            builder.field(Names.TREE_PRESISION, DistanceUnit.METERS.toString(fieldType().precisionInMeters()));
-        } else if (includeDefaults && fieldType().treeLevels() == 0) { // defaults only make sense if tree levels are not specified
-            builder.field(Names.TREE_PRESISION, DistanceUnit.METERS.toString(50));
-        }
-        if (includeDefaults || fieldType().strategyName().equals(Defaults.STRATEGY) == false) {
-            builder.field(Names.STRATEGY, fieldType().strategyName());
-        }
-        if (includeDefaults || fieldType().distanceErrorPct() != fieldType().defaultDistanceErrorPct) {
-            builder.field(Names.DISTANCE_ERROR_PCT, fieldType().distanceErrorPct());
-        }
-        if (includeDefaults || fieldType().orientation() != Defaults.ORIENTATION) {
-            builder.field(Names.ORIENTATION, fieldType().orientation());
-        }
-        if (fieldType().strategyName().equals(SpatialStrategy.TERM.getStrategyName())) {
-            // For TERMs strategy the defaults for points only change to true
-            if (includeDefaults || fieldType().pointsOnly() != true) {
-                builder.field(Names.STRATEGY_POINTS_ONLY, fieldType().pointsOnly());
+                    fieldType().typeName());
+            }
+            context.addIgnoredField(fieldType().name());
+        }
+    }
+
+    private void indexShape(ParseContext context, Object luceneShape) {
+        if (luceneShape instanceof GeoPoint) {
```

Comment:
having a root class doesn't make much sense from a Lucene perspective since you're not expected to manipulate them in a generic way. Maybe we could have a wrapper on the Elasticsearch side to add type safety that would look something like this? Would it address your concern?

```java
public class GeoShape {

  // exactly one of them is non-null
  public final double[] point;
  public final Line line;
  public final Polygon polygon;
  // other types we want to support

  public GeoShape(double[] point) {
    this.point = point;
    this.line = null;
    this.polygon = null;
  }

  public GeoShape(Line line) {
    this.point = null;
    this.line = line;
    this.polygon = null;
  }

  public GeoShape(Polygon polygon) {
    this.point = null;
    this.line = null;
    this.polygon = polygon;
  }
}
```

Reply:
I think this will just replace this huge `if (shape instanceof ... )` with `if (shape != null)` switch.

Reply:
Indeed. I was trying to see whether that would address your concern that there is no root geometry class.

Reply:
@imotov perhaps this can be resolved in a future PR by using the coming Geo API?

Reply:
:+1: